coap/adapter.go

```diff
@@ -1,61 +1,116 @@
-package adapter
+package coap
 
 import (
-	"encoding/json"
-	"log"
-	"net"
+	"errors"
+	"sync"
+	"time"
 
-	"github.com/dustin/go-coap"
 	"github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/coap/nats"
 	broker "github.com/nats-io/go-nats"
-	"go.uber.org/zap"
 )
 
-const protocol string = "coap"
+const (
+	key     string = "key"
+	channel string = "id"
+)
 
-type Observer struct {
-	conn    *net.UDPConn
-	addr    *net.UDPAddr
-	message *coap.Message
-}
+var (
+	errBadRequest    = errors.New("bad request")
+	errBadOption     = errors.New("bad option")
+	errEntryNotFound = errors.New("observer entry not founds")
+)
 
-type CoAPAdapter struct {
-	obsMap map[string][]Observer
-	logger *zap.Logger
-	pub    mainflux.MessagePublisher
+// AdapterService struct represents CoAP adapter service implementation.
+type adapterService struct {
+	pubsub nats.Service
+	subs   map[string]nats.Observer
+	mu     sync.Mutex
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
-		logger: logger,
-		pub:    pub,
-		obsMap: make(map[string][]Observer),
+// New creates new CoAP adapter service struct.
+func New(pubsub nats.Service) Service {
+	return &adapterService{
+		pubsub: pubsub,
+		subs:   make(map[string]nats.Observer),
+		mu:     sync.Mutex{},
 	}
+}
 
-	return ca
+func (svc *adapterService) Publish(msg mainflux.RawMessage) error {
+	if err := svc.pubsub.Publish(msg); err != nil {
+		switch err {
+		case broker.ErrConnectionClosed, broker.ErrInvalidConnection:
+			return ErrFailedConnection
+		default:
+			return ErrFailedMessagePublish
+		}
+	}
+	return nil
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+func (svc *adapterService) Subscribe(chanID, clientID string, ch chan mainflux.RawMessage) error {
```

Comment:
Technically, you're not unsubscribing but replacing the connection, as far as I understand (with resource cleanup). Since that's the case, I would consider making it explicit through function naming (and their implementation), and rename the `Unsubscribe` to `CreateOrReplaceConnection`.